kernel/arch/i386/boot/bootparse.cpp

```diff
@@ -116,10 +123,166 @@ void px_parse_multiboot2(void *info)
+            (uint32_t)(entry->length >> 32) & 0xff, (uint32_t)(entry->length & UINT32_MAX),
+            // Have to do this because the IDs aren't in an order that can be
+            // mapped into an array like Multiboot. The last two types start at
+            // 0x1000, so we have to do some ternary operator magic.
+            (entry->type < 6 ? mmap_types[entry->type]
+                             : (entry->type == STIVALE2_MMAP_BOOTLOADER_RECLAIMABLE ? "Bootloader"
+                                                                                    : "Kernel & Modules")));
+    }
+}
+
+void px_parse_stivale2(void *info)
+{
+    auto fixed = (struct stivale2_struct*)info;
+    for (uintptr_t page = ((uintptr_t)info & PAGE_ALIGN) + PAGE_SIZE;
+         page <= ((uintptr_t)info & PAGE_ALIGN);
+         page += PAGE_SIZE) {
+        px_rs232_printf("Mapping bootinfo at 0x%08x\n", page);
+        px_map_kernel_page(VADDR(page), page);
+    }
+    // Walk the list of tags in the header
+    auto tag = (struct stivale2_tag*)(fixed->tags);
+    while (tag)
+    {
+        // Map in each tag since Stivale2 doesn't give us a total size like Multiboot does.
+        uintptr_t page = ((uintptr_t)tag & PAGE_ALIGN);
+        px_map_kernel_page(VADDR(page), page);
+        // Follows the tag list order in stivale2.h
+        switch(tag->identifier)
+        {
+            case STIVALE2_STRUCT_TAG_CMDLINE_ID:
+            {
+                auto cmdline = (struct stivale2_struct_tag_cmdline*)tag;
+                px_rs232_printf("Stivale2 cmdline: '%s'\n", (const char *)cmdline->cmdline);
+                break;
+            }
+            case STIVALE2_STRUCT_TAG_MEMMAP_ID:
+            {
+                auto memmap = (struct stivale2_struct_tag_memmap*)tag;
+                px_rs232_printf("Stivale2 memory map:\n");
+                print_stivale2_mmap(memmap);
+                break;
+            }
+            case STIVALE2_STRUCT_TAG_FRAMEBUFFER_ID:
+            {
+                auto framebuffer = (struct stivale2_struct_tag_framebuffer*)tag;
+                px_rs232_printf("Stivale2 framebuffer:\n");
+                px_rs232_printf("  Address: 0x%08X", framebuffer->framebuffer_addr);
+                px_rs232_printf("  Resolution: %ix%ix%i\n",
+                    framebuffer->framebuffer_width,
+                    framebuffer->framebuffer_height,
+                    (framebuffer->framebuffer_bpp * 8));
+                break;
+            }
+            case STIVALE2_STRUCT_TAG_FB_MTRR_ID:
+            {
+                px_rs232_printf("  Framebuffer has MTRR\n");
+                break;
+            }
+            case STIVALE2_STRUCT_TAG_MODULES_ID:
+            {
+                auto modules = (struct stivale2_struct_tag_modules*)tag;
+                for (uint64_t i = 0; i < modules->module_count; i++) {
+                    stivale2_module mod = modules->modules[i];
```

Comment:
Any reason you copy the module structure vs. referencing it in memory?

Reply:
None other than it was simple to index the array. I will change it to a pointer.

Reply:
Resolved in 4122625